include/fixed_containers/unified_interface_red_black_tree.hpp

```diff
@@ -0,0 +1,201 @@
+          : tree_{nullptr}
+          , current_index_{CAPACITY}
+        {
+        }
+
+        constexpr PairProvider(const PairProvider&) = default;
+        constexpr PairProvider(PairProvider&&) noexcept = default;
+        constexpr PairProvider& operator=(const PairProvider&) = default;
+        constexpr PairProvider& operator=(PairProvider&&) noexcept = default;
+
+        // https://github.com/llvm/llvm-project/issues/62555
+        template <bool IS_CONST_2>
+        constexpr PairProvider(const PairProvider<IS_CONST_2>& m) noexcept
+            requires(IS_CONST and !IS_CONST_2)
+          : PairProvider{m.tree_, {.parent = NULL_INDEX, .i = m.current_index_, .is_left_child = false}}
+        {
+        }
+
+        constexpr void advance() noexcept
+        {
+            if (current_index_ == NULL_INDEX)
+            {
+                current_index_ = tree_->index_of_min_at().i;
+            }
+            else
+            {
+                current_index_ = tree_->index_of_successor_at(current_index_);
+                current_index_ = replace_null_index_with_max_size_for_end_iterator(current_index_);
+            }
+        }
+        constexpr void recede() noexcept
+        {
+            if (current_index_ == MAXIMUM_SIZE)
+            {
+                current_index_ = tree_->index_of_max_at().i;
+            }
+            else
+            {
+                current_index_ = tree_->index_of_predecessor_at(current_index_);
+            }
+        }
+
+        constexpr std::conditional_t<IS_CONST, const_reference, reference> get() const noexcept
+        {
+            fixed_red_black_tree_detail::RedBlackTreeNodeView node = tree_->node_at(current_index_);
+            return {node.key(), node.value()};
+        }
+
+        template <bool IS_CONST2>
+        constexpr bool operator==(const PairProvider<IS_CONST2>& other) const noexcept
+        {
+            return tree_ == other.tree_ && current_index_ == other.current_index_;
+        }
+    };
+
+    constexpr OpaqueIndexType opaque_index_at(const K& k) const
+    {
+        return this->index_of_node_with_parent(k);
+    }
+
+    constexpr OpaqueIndexType begin() const
+    {
+        return this->index_of_min_at();
+    }
+
+    constexpr OpaqueIndexType end() const
+    {
+        return {.parent = NULL_INDEX, .i = CAPACITY, .is_left_child = false};
+    }
+
+    constexpr bool exists(const OpaqueIndexType& i) const
+    {
+        // search will terminate with i being NULL if the key is not found
+        return i.i != NULL_INDEX;
+    }
+
+    constexpr const V& value(const OpaqueIndexType& i) const
+    {
+        // no safety checks
+        return this->node_at(i.i).value();
+    }
+
+    constexpr V& value(const OpaqueIndexType& i)
+    {
+        return this->node_at(i.i).value();
+    }
+
+    template <typename... Args>
+    constexpr OpaqueIndexType emplace(const OpaqueIndexType& i, const K& key, Args&&... args)
+    {
```

Comment:
Since the class follows `std::map` API names, this should be `try_emplace()` (instead of `emplace()`), as it separates the key and value args.